src/main/scala/com/mozilla/telemetry/stats/MannWhitneyU.scala

```diff
@@ -0,0 +1,120 @@
+/*
+ This Source Code Form is subject to the terms of the Mozilla Public
+ License, v. 2.0. If a copy of the MPL was not distributed with this
+ file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+package com.mozilla.telemetry.stats
+
+
+// Constants
+
+object MannWhitneyU {
+
+  def erf(x: Double): Double =  {
+    // Constants.
+    val a1: Double = 0.254829592
+    val a2: Double = -0.284496736
+    val a3: Double = 1.421413741
+    val a4: Double = -1.453152027
+    val a5: Double = 1.061405429
+    val p: Double = 0.3275911
+
+    if (x == 0) {
+      return 0.0
+    }
+
+    val sign: Int = if (x < 0) -1 else 1
+    val absx: Double = math.abs(x)
+    val t: Double = 1.0 / (1.0 + p * absx);
+    val y: Double = 1.0 - (((((a5 * t + a4) * t) + a3) * t + a2) * t + a1) * t * math.exp(-x * x);
+    sign * y
+  }
+
+  def erfc(x: Double): Double = {
+    1.0 - erf(x)
+  }
+
+  def ndtr(a: Double): Double = {
+    /*
+    Returns the area under the Gaussian probability density function,
+    integrated from minus infinity to x.
+
+    See: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ndtr.html#scipy.special.ndtr
+
+    */
+    val sqrth: Double = math.sqrt(2) / 2
+    val x: Double = a * sqrth
+    val z: Double = math.abs(x)
+    var y: Double = 0.0
+
+    if (z < sqrth) {
+        y = 0.5 + 0.5 * erf(x)
+    } else {
+        y = 0.5 * erfc(z)
+        if (x > 0) {
+            y = 1.0 - y
+        }
+    }
+    return y
+  }
+
+  def tie_correct(sample: Map[Int, Int]): Double = {
+    var tc: Double = 0.0
+    val n = sample.foldLeft[Int](0)(_ + _._2)
+
+    if (n < 2) {
+      return 1.0
+    }
+
+    tc = sample.foldLeft[Double](tc)((sum, i) => sum + math.pow(i._2, 3) - i._2)
+    tc = 1 - tc / (math.pow(n, 3) - n)
+    tc
+  }
+
+  def rank(sample: Map[Int, Int]): Map[Int, Double] = {
+    var rank: Int = 1
```

Comment:
See if you can make this function just as performant without the variable.

Reply:
I tried using `foldLeft` instead of the `foreach` and ran into a problem. I'm not seeing how we build the new `ranks` map which depends on the `rank` variable w/o the variable. I was thinking `foldLeft` takes a seed value for the summation and maybe that's where you were leading me? But since `foldLeft` just does an iteration under the hood it doesn't seem more performant either way.